Dockerfile

```diff
@@ -1,31 +1,39 @@
-FROM python:3.8-slim-buster as builder
+FROM python:3.9-slim as base
 
 ENV LANG=C.UTF-8 \
     DEBIAN_FRONTEND=noninteractive \
     PIP_NO_CACHE_DIR=true
 
-RUN apt-get update && apt-get install --no-install-recommends -qq -y gcc g++ && apt-get clean autoclean \
-  && apt-get autoremove -y \
-  && rm -rf /var/lib/apt/lists/* \
-  && rm -f /var/cache/apt/archives/*.deb
+RUN apt-get update && apt-get install -y --no-install-recommends -qq gcc=4:10.2.1-1 libffi-dev=3.3-6 g++=4:10.2.1-1 git=1:2.30.2-1 curl=7.74.0-1.3+deb11u1 \
+ && apt-get clean \
+ && rm -rf /var/lib/apt/lists/*
 
-COPY requirements.txt .
-RUN pip install --upgrade pip && pip install --trusted-host pypi.python.org -r requirements.txt
+WORKDIR /app
 
-FROM python:3.8-slim-buster as production
+FROM base as builder
 
-ENV PYTHONDONTWRITEBYTECODE 1
-ENV PYTHONUNBUFFERED 1
+ENV POETRY_VERSION=1.1.13
+RUN pip install --no-cache-dir poetry==$POETRY_VERSION
 
-WORKDIR /app
+COPY pyproject.toml poetry.lock ./
+RUN python -m venv --copies /venv
+
+RUN . /venv/bin/activate && poetry install --no-dev --no-root
+
+FROM base as production
+
+COPY --from=builder /venv /venv
 
 RUN mkdir -p /var/www && chown www-data /var/www && \
-    apt-get update && apt-get install --no-install-recommends -qq -y curl && \
     apt-get clean && find /var/lib/apt/lists/ -type f -delete && \
```

Comment:
- don't we need `curl`? it is used in the healthcheck
- there is still `apt-get clean' but `apt-get update && apt-get install` was removed

Reply:
Installing curl is going in "base" stage
clean and lists removed, because executed on base stage